### PR TITLE
Add Python 3.12 to supported versions and test against on GitHub Actions; upgrade versioneer

### PR DESCRIPTION
**Pull request by yarikoptic**

### CODE REVIEW

**Comment by yarikoptic** [comment]
aiohttp needs a new release 3.9 with 3.12 support: https://github.com/aio-libs/aiohttp/issues/7675

**Comment by yarikoptic** [comment]
@jwodder do you think we should open a can of "allow beta/alpha releases" for aiohttp to allow 3.12 here?

**Comment by jwodder** [comment]
@yarikoptic We could add prereleases of aiohttp to the packages tested in the "dev-deps" job [here](https://github.com/dandi/dandi-cli/blob/cb86c9cb3fad0b8735ad86eb4ba43569d4a0a425/.github/workflows/test.yml#L70), but if you're suggesting that installing dandi-cli should install a prerelease of aiohttp under normal circumstances, the only way to do that would be for dandi-cli to require a specific prerelease version of aiohttp as the only version to install, which seems like a bad idea.

**Comment by yarikoptic** [comment]
aiohttp  3.9.0 is out with 3.12 support.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -36,6 +36,7 @@ jobs:
           - 3.9
           - '3.10'  # Needs quotes so YAML doesn't think it's 3.1
           - '3.11'
+          - '3.12.0-rc.3'
```
Comment (yarikoptic) [inline .github/workflows/test.yml:39]: according to https://github.com/actions/setup-python/issues/736 3.12 was released so I guess we can just use it

```suggestion
          - '3.12'
```

**Comment by jwodder** [comment]
@yarikoptic FYI: https://github.com/dandi/dandi-cli/issues/1353

File: .github/workflows/test.yml (hunk 1 above)
Comment (yarikoptic) [reply .github/workflows/test.yml:39]: ha -- forgot to commit... well -- will do locally since anyways need to add more ignores

**Comment by yarikoptic** [comment]
<details>
<summary>committed to use released 3.12 so may be that somehow would help to avoid triggering nested deprecation warning ... if not -- I will add ignore for that one too above the one which triggers it.
</summary> 

```shell
ERROR: while parsing the following warning configuration:
15

16
  ignore:.*Value with data type .* is being converted:hdmf.build.warnings.DtypeConversionWarning
17

18
This error occurred:
19

20
Traceback (most recent call last):
21
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/_pytest/config/__init__.py", line 1761, in parse_warning_filter
22
Warning: ory: Type[Warning] = _resolve_warning_category(category_)
23
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
24
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/_pytest/config/__init__.py", line 1799, in _resolve_warning_category
25
    m = __import__(module, None, None, [klass])
26
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
27
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/hdmf/__init__.py", line 2, in <module>
28
    from .backends.hdf5.h5_utils import H5Dataset, H5RegionSlicer
29
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/hdmf/backends/__init__.py", line 1, in <module>
30
    from . import hdf5
31
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/hdmf/backends/hdf5/__init__.py", line 1, in <module>
32
    from . import h5_utils, h5tools
33
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/hdmf/backends/hdf5/h5_utils.py", line 22, in <module>
34
    from ...region import RegionSlicer
35
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/hdmf/region.py", line 4, in <module>
36
    from .container import Data, DataRegion
37
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/hdmf/container.py", line 10, in <module>
38
    import pandas as pd
39
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/pandas/__init__.py", line 46, in <module>
40
    from pandas.core.api import (
41
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/pandas/core/api.py", line 1, in <module>
42
    from pandas._libs import (
43
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/pandas/_libs/__init__.py", line 18, in <module>
44
    from pandas._libs.interval import Interval
45
  File "interval.pyx", line 1, in init pandas._libs.interval
46
  File "hashtable.pyx", line 1, in init pandas._libs.hashtable
47
  File "missing.pyx", line 1, in init pandas._libs.missing
48
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/pandas/_libs/tslibs/__init__.py", line 39, in <module>
49
    from pandas._libs.tslibs.conversion import localize_pydatetime
50
  File "conversion.pyx", line 1, in init pandas._libs.tslibs.conversion
51
  File "offsets.pyx", line 1, in init pandas._libs.tslibs.offsets
52
  File "timestamps.pyx", line 1, in init pandas._libs.tslibs.timestamps
53
  File "timedeltas.pyx", line 1, in init pandas._libs.tslibs.timedeltas
54
  File "timezones.pyx", line 24, in init pandas._libs.tslibs.timezones
55
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/dateutil/tz/__init__.py", line 2, in <module>
56
    from .tz import *
57
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/dateutil/tz/tz.py", line 37, in <module>
58
    EPOCH = datetime.datetime.utcfromtimestamp(0)
59
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
60
DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
```
</details>

**Comment by yarikoptic** [comment]
green besides codecov and some odd on windows 3.9 which can't relate:

```
________________________ test_IteratorWithAggregation _________________________
dandi\support\tests\test_iterators.py:72: in test_IteratorWithAggregation
    assert got in ([], [0])
E   assert [0, 1] in ([], [0])

---------- coverage: platform win32, python 3.9.13-final-0 -----------
```

so let's proceed.

**Pull request merged**